services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/team/TeamManagerImpl.java

```diff
@@ -484,41 +490,63 @@ private boolean hasUnmetAccessRequirements(UserInfo memberUserInfo, String teamI
 		return restrictionInformationManager.getRestrictionInformation(memberUserInfo, request).getHasUnmetAccessRequirement();
 	}
 
+	private AuthorizationStatus checkCanAddOther(String teamId, boolean amTeamAdmin, String principalId, long now) {
+		// the member to be added is someone other than me
+		if (!amTeamAdmin) return UNAUTHORIZED_ADD_TEAM_MEMBER_MUST_BE_TEAM_MANAGER; // can't add someone unless I'm a Team administrator
+		// can't add someone unless they are asking to be added
+		long openRequestCount = membershipRequestDAO.getOpenByTeamAndRequesterCount(Long.parseLong(teamId), Long.parseLong(principalId), now);
+		return openRequestCount>0L  ? AUTHORIZED_ADD_TEAM_MEMBER : UNAUTHORIZED_ADD_TEAM_MEMBER_MUST_HAVE_REQUEST;
+	}
+
+	private AuthorizationStatus checkCanAddSelf(String teamId, boolean amTeamAdmin, String principalId, long now) {
+		// trying to add myself to Team.
+		if (amTeamAdmin) return AUTHORIZED_ADD_TEAM_MEMBER;
+		// if the team is open, I can join
+		Team team = teamDAO.get(teamId);
+		if (team.getCanPublicJoin()!=null && team.getCanPublicJoin()==true) return AUTHORIZED_ADD_TEAM_MEMBER;
+		// if I'm not a team admin and the team is not open, then I need to have an open invitation
+		long openInvitationCount = membershipInvitationDAO.getOpenByTeamAndUserCount(Long.parseLong(teamId), Long.parseLong(principalId), now);
+		return openInvitationCount>0L ? AUTHORIZED_ADD_TEAM_MEMBER : UNAUTHORIZED_ADD_TEAM_MEMBER_MUST_HAVE_INVITATION;
+	}
+
+
 	/**
 	 * Either:
 		principalId is self and membership invitation has been extended (and not yet accepted), or
     	principalId is self and have MEMBERSHIP permission on Team, or
     	have MEMBERSHIP permission on Team and membership request has been created (but not yet accepted) for principalId
 	 * @param userInfo
 	 * @param teamId the ID of the team
-	 * @param principalId the ID of the one to be added to the team
 	 * @return
 	 */
 	public AuthorizationStatus canAddTeamMember(UserInfo userInfo, String teamId, UserInfo principalUserInfo, boolean alreadyInTeam) throws NotFoundException {
+		// admin can always accept membership
 		if (userInfo.isAdmin()) return AUTHORIZED_ADD_TEAM_MEMBER;
-		if (hasUnmetAccessRequirements(principalUserInfo, teamId)) return UNAUTHORIZED_ADD_TEAM_MEMBER_UNMET_AR;
+		if (alreadyInTeam) return AUTHORIZED_ADD_TEAM_MEMBER;
```

Comment:
What are the implications of returning 'true' when the member is already in the team?